cpp/src/arrow/compute/kernels/aggregate_test.cc

```diff
@@ -308,5 +312,59 @@ TYPED_TEST(TestRandomNumericCountKernel, RandomArrayCount) {
   }
 }
 
+///
+/// Min / Max
+///
+
+template <typename ArrowType>
+class TestNumericMinMaxKernel : public ComputeFixture, public TestBase {
+  using Traits = TypeTraits<ArrowType>;
+  using ArrayType = typename Traits::ArrayType;
+  using c_type = typename ArrayType::value_type;
+  using ScalarType = typename Traits::ScalarType;
+
+ public:
+  template <typename T>
+  void AssertMinMaxIs(std::string array_json, T expected_min, T expected_max,
+                      const MinMaxOptions& options) {
+    auto array = ArrayFromJSON(Traits::type_singleton(), array_json);
+    Datum out, out_min, out_max;
+    ASSERT_OK(MinMax(&this->ctx_, options, *array, &out));
+
+    ASSERT_TRUE(out.is_collection());
+    auto col = out.collection();
+
+    out_min = col[0];
+    ASSERT_TRUE(out_min.is_scalar());
+    auto min = checked_pointer_cast<ScalarType>(out_min.scalar());
+    ASSERT_EQ(min->value, static_cast<c_type>(expected_min));
+
+    out_max = col[1];
+    ASSERT_TRUE(out_max.is_scalar());
+    auto max = checked_pointer_cast<ScalarType>(out_max.scalar());
+    ASSERT_EQ(max->value, static_cast<c_type>(expected_max));
+  }
+};
+
+template <typename ArrowType>
+class TestFloatingMinMaxKernel : public TestNumericMinMaxKernel<ArrowType> {};
+
+TYPED_TEST_CASE(TestNumericMinMaxKernel, IntegralArrowTypes);
+TYPED_TEST(TestNumericMinMaxKernel, Basics) {
+  MinMaxOptions options;
+  this->AssertMinMaxIs("[5, 1, 2, 3, 4]", 1, 5, options);
+  this->AssertMinMaxIs("[5, null, 2, 3, 4]", 2, 5, options);
+}
+
+TYPED_TEST_CASE(TestFloatingMinMaxKernel, RealArrowTypes);
+TYPED_TEST(TestFloatingMinMaxKernel, Floats) {
+  MinMaxOptions options;
+  this->AssertMinMaxIs("[5, 1, 2, 3, 4]", 1, 5, options);
+  this->AssertMinMaxIs("[5, null, 2, 3, 4]", 2, 5, options);
+  // this->AssertMinMaxIs("[5, Inf, 2, 3, 4]", 2.0, INFINITY, options);
```

Comment:
For this line, I run into the following problem:

```
../src/arrow/compute/kernels/aggregate_test.cc: In instantiation of 'void arrow::compute::TestFloatingMinMaxKernel_Floats_Test<gtest_TypeParam_>::TestBody() [with gtest_TypeParam_ = arrow::DoubleType]':
../src/arrow/array/builder_binary.h:398:20:   required from here
../src/arrow/compute/kernels/aggregate_test.cc:366:3: error: no matching function for call to 'arrow::compute::TestFloatingMinMaxKernel_Floats_Test<arrow::DoubleType>::AssertMinMaxIs(const char [19], float, int, arrow::compute::MinMaxOptions&)'
   this->AssertMinMaxIs("[5, -Inf, 2, 3, 4]", -INFINITY, 5, options);
   ^~~~
../src/arrow/compute/kernels/aggregate_test.cc:329:8: note: candidate: template<class T> void arrow::compute::TestNumericMinMaxKernel<ArrowType>::AssertMinMaxIs(std::__cxx11::string, T, T, const arrow::compute::MinMaxOptions&) [with T = T; ArrowType = arrow::DoubleType]
   void AssertMinMaxIs(std::string array_json, T expected_min, T expected_max, const MinMaxOptions& options) {
        ^~~~~~~~~~~~~~
../src/arrow/compute/kernels/aggregate_test.cc:329:8: note:   template argument deduction/substitution failed:
../src/arrow/compute/kernels/aggregate_test.cc:366:3: note:   deduced conflicting types for parameter 'T' ('float' and 'int')
   this->AssertMinMaxIs("[5, -Inf, 2, 3, 4]", -INFINITY, 5, options);
```

so as long as both arguments are integers (eg the line above), it seems fine to interpret them as floats or doubles. 
Is there a way that I can use one of the template variables of above to already cast this to the correct type?

Reply:
The simplest fix is to rewrite `AssertMinMaxIs` with separate template arguments for min and max type.
```c++
  template <typename Min, typename Max>
  void AssertMinMaxIs(std::string array_json, Min expected_min, Max expected_max, 
```

Reply:
Thanks, that works!